include/pybind11/pytypes.h

```diff
@@ -375,31 +388,101 @@ PYBIND11_NAMESPACE_END(detail)
 /// python).
 class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
-    /// Constructs a new exception from the current Python error indicator, if any.  The current
-    /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    /// Fetches the current Python exception (using PyErr_Fetch()), which will clear the
+    /// current Python error indicator.
+    error_already_set() : std::runtime_error("") {
+        if (!PyErr_Occurred()) {
+            pybind11_fail("Internal error: pybind11::detail::error_already_set called while "
+                          "Python error indicator not set.");
+        }
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
     }
 
     error_already_set(const error_already_set &) = default;
```

Comment:
std::runtime copy ctor is noexcept so I am unsure if we should specify that here. The only real issue is the allocation of the new std::string could throw I suppose.